pkg/crc/machine/bundle/metadata.go

```diff
@@ -263,21 +275,40 @@ func getBundleDownloadInfo(preset preset.Preset) (*download.RemoteFile, error) {
 	return downloadInfo, nil
 }
 
-func Download(preset preset.Preset) error {
+func DownloadDefault(preset crcPreset.Preset) (string, error) {
 	downloadInfo, err := getBundleDownloadInfo(preset)
 	if err != nil {
-		return err
+		return "", err
 	}
 	bundleNameFromURI, err := downloadInfo.GetSourceFilename()
 	if err != nil {
-		return err
+		return "", err
 	}
 	if constants.GetDefaultBundle(preset) != bundleNameFromURI {
-		return fmt.Errorf("expected %s but found %s in Makefile", bundleNameFromURI, constants.GetDefaultBundle(preset))
-	}
-	if _, err := downloadInfo.Download(constants.GetDefaultBundlePath(preset), 0664); err != nil {
-		return err
+		return "", fmt.Errorf("expected %s but found %s in Makefile", bundleNameFromURI, constants.GetDefaultBundle(preset))
 	}
 
-	return nil
+	return downloadInfo.Download(constants.GetDefaultBundlePath(preset), 0664)
+}
+
+func Download(preset crcPreset.Preset, bundleURI string) (string, error) {
+	// If we are asked to download
+	// ~/.crc/cache/crc_podman_libvirt_4.1.1.crcbundle, this means we want
+	// are downloading the default bundle for this release. This uses a
+	// different codepath from user-specified URIs as for the default
+	// bundles, their sha256sums are known and can be checked.
+	if bundleURI == constants.GetDefaultBundlePath(preset) {
+		if preset == crcPreset.OpenShift {
+			return DownloadDefault(preset)
+		}
+		return image.PullBundle(preset, "")
+	}
+	switch {
+	case strings.HasPrefix(bundleURI, "http://"), strings.HasPrefix(bundleURI, "https://"):
+		return download.Download(bundleURI, constants.MachineCacheDir, 0644, nil)
+	case strings.HasPrefix(bundleURI, "docker://"):
+		return image.PullBundle(preset, strings.TrimPrefix(bundleURI, "docker:"))
```

Comment:
... as this also `image.PullBundle` sounds more like `registry` or even `docker.DownloadBundle`.

Not sure if the package separation follows logical naming, especially as 'download' is now a scheme-specific solution to 'download'.

Reply:
When developing some new code, it's useful to cleanup preexisting code which is related to the code we are changing while this code path is fresh in our minds :)